### PR TITLE
fix(packages): Add label for private and internal package and fix composor package source permission check

### PR DESCRIPTION
**Pull request by lunny**
- Add permission checks for Composer package source links

- Add private/internal visibility labels for packages, similar to repository visibility labels

image

- Add a link to change package visibility

image

- Update link package descriptions

image

### CODE REVIEW

**Comment by bircni** [comment]
1. **`HasAnyUnitAccess()` may hide source URLs for repos with public-unit-only access.** The permission guard in `routers/api/packages/composer/api.go:125` uses `permission.HasAnyUnitAccess()`, which only considers explicit (`unitsMode`) and direct (`AccessMode`) access — not the "everyone" or "anonymous" access modes. A repo where only some units are publicly accessible (but the overall `AccessMode` is `None`) would pass `HasAnyUnitAccessOrPublicAccess()` but fail `HasAnyUnitAccess()`, incorrectly hiding the source URL from legitimate readers. The TODO comment at `models/perm/access/repo_permission.go:50` explicitly calls this out. The fix should use `permission.HasAnyUnitAccessOrPublicAccess()` to match the access semantics used elsewhere in the codebase (e.g. `routers/api/v1/api.go:211`).

2. **`Package.IsInternal` is mislabeled as "Private" in the visibility badge.** In `templates/package/shared/visibility_badge.tmpl:1–3`, `Package.IsInternal` triggers the "Private" badge (`repo.desc.private`). However, `IsInternal` denotes Gitea-internal packages (CI/CD artefacts, container blob manifests — see `services/packages/packages.go:439,454`), which are a completely different concept from user-restricted private packages. If an internal package somehow surfaces in a listing, the "Private" label would be misleading. Since `repo.desc.internal` already exists in locales, the first branch should use that key, or the flag should be excluded from the badge altogether (internal packages are already filtered from public views).

3. **The `WebVisibilityBadge` test only covers the no-badge path.** `tests/integration/api_packages_composer_test.go:286–300` asserts that zero badge elements exist for a public user's packages, but never asserts that a badge *does* appear when the owner is private or limited. The test passes vacuously for the presence logic. Add a sub-case that creates/uses a package owned by a private org or user and confirms the badge renders (e.g. check `assert.Equal(t, 1, doc.Find(".ui.basic.label").Length())`).

4. **A single DB error in `createPackageMetadataResponse` aborts the whole package response.** At `routers/api/packages/composer/api.go:122`, if `GetDoerRepoPermission` returns an error for any descriptor the function immediately propagates it, returning HTTP 500 for the entire `PackageMetadata` response. Since this call is non-critical (visibility of the source URL, not the package itself), consider logging the error and omitting the source field rather than failing the entire request, which degrades the experience for all packages in the response.

5. **`cleanup_rules/preview.tmpl` passes `.Owner` without verifying it is always populated in that template's context.** `templates/package/shared/cleanup_rules/preview.tmpl:22` now calls `dict "Package" .Package "Owner" .Owner` inside a `{{range .VersionsToRemove}}`. The range element is a `*packages_model.PackageDescriptor`, which does carry an `Owner` field populated via `GetPackageDescriptor`. However, there is no test covering the cleanup-rules preview path with the badge, so a regression (e.g. a future refactor that uses a lighter struct without Owner) would be silent. Adding a test or at minimum a nil-guard in the template (`{{if .Owner}}`) would be prudent.

_Written by Claude_

File: routers/api/packages/composer/api.go [hunk 1]
```diff
@@ -91,7 +93,7 @@ type Source struct {
 	Reference string `json:"reference"`
 }
 
-func createPackageMetadataResponse(registryURL string, pds []*packages_model.PackageDescriptor) *PackageMetadataResponse {
+func createPackageMetadataResponse(ctx *context.Context, registryURL string, pds []*packages_model.PackageDescriptor) (*PackageMetadataResponse, error) {
```
[the remark this answers is not in the document]
Comment (lunny) [reply routers/api/packages/composer/api.go]: It's in router layer and it needs `ctx.Doer`.

File: tests/integration/api_packages_composer_test.go [hunk 2]
```diff
@@ -243,5 +244,49 @@ func TestPackageComposer(t *testing.T) {
...
+			AddBasicAuth(user.Name)
+		listResp := MakeRequest(t, listReq, http.StatusOK)
+		listDoc := NewHTMLParser(t, bytes.NewReader(listResp.Body.Bytes()))
+		assert.Equal(t, 0, listDoc.Find(".item-title .ui.basic.label").Length())
+
+		viewReq := NewRequest(t, "GET", fmt.Sprintf("/%s/-/packages/composer/%s/%s", user.Name, neturl.PathEscape(packageName), neturl.PathEscape(packageVersion))).
+			AddBasicAuth(user.Name)
+		viewResp := MakeRequest(t, viewReq, http.StatusOK)
+		viewDoc := NewHTMLParser(t, bytes.NewReader(viewResp.Body.Bytes()))
+		assert.Equal(t, 0, viewDoc.Find(".issue-title-header .ui.basic.label").Length())
```
[the remark this answers is not in the document]
Comment (lunny) [reply tests/integration/api_packages_composer_test.go:291]: This has been resolved by 2d97484be9746c5aa759fe4d536250f0a83fe996

**Comment by lunny** [comment]
> 1. **`HasAnyUnitAccess()` may hide source URLs for repos with public-unit-only access.** The permission guard in `routers/api/packages/composer/api.go:125` uses `permission.HasAnyUnitAccess()`, which only considers explicit (`unitsMode`) and direct (`AccessMode`) access — not the "everyone" or "anonymous" access modes. A repo where only some units are publicly accessible (but the overall `AccessMode` is `None`) would pass `HasAnyUnitAccessOrPublicAccess()` but fail `HasAnyUnitAccess()`, incorrectly hiding the source URL from legitimate readers. The TODO comment at `models/perm/access/repo_permission.go:50` explicitly calls this out. The fix should use `permission.HasAnyUnitAccessOrPublicAccess()` to match the access semantics used elsewhere in the codebase (e.g. `routers/api/v1/api.go:211`).
> 2. **`Package.IsInternal` is mislabeled as "Private" in the visibility badge.** In `templates/package/shared/visibility_badge.tmpl:1–3`, `Package.IsInternal` triggers the "Private" badge (`repo.desc.private`). However, `IsInternal` denotes Gitea-internal packages (CI/CD artefacts, container blob manifests — see `services/packages/packages.go:439,454`), which are a completely different concept from user-restricted private packages. If an internal package somehow surfaces in a listing, the "Private" label would be misleading. Since `repo.desc.internal` already exists in locales, the first branch should use that key, or the flag should be excluded from the badge altogether (internal packages are already filtered from public views).
> 3. **The `WebVisibilityBadge` test only covers the no-badge path.** `tests/integration/api_packages_composer_test.go:286–300` asserts that zero badge elements exist for a public user's packages, but never asserts that a badge _does_ appear when the owner is private or limited. The test passes vacuously for the presence logic. Add a sub-case that creates/uses a package owned by a private org or user and confirms the badge renders (e.g. check `assert.Equal(t, 1, doc.Find(".ui.basic.label").Length())`).
> 4. **A single DB error in `createPackageMetadataResponse` aborts the whole package response.** At `routers/api/packages/composer/api.go:122`, if `GetDoerRepoPermission` returns an error for any descriptor the function immediately propagates it, returning HTTP 500 for the entire `PackageMetadata` response. Since this call is non-critical (visibility of the source URL, not the package itself), consider logging the error and omitting the source field rather than failing the entire request, which degrades the experience for all packages in the response.
> 5. **`cleanup_rules/preview.tmpl` passes `.Owner` without verifying it is always populated in that template's context.** `templates/package/shared/cleanup_rules/preview.tmpl:22` now calls `dict "Package" .Package "Owner" .Owner` inside a `{{range .VersionsToRemove}}`. The range element is a `*packages_model.PackageDescriptor`, which does carry an `Owner` field populated via `GetPackageDescriptor`. However, there is no test covering the cleanup-rules preview path with the badge, so a regression (e.g. a future refactor that uses a lighter struct without Owner) would be silent. Adding a test or at minimum a nil-guard in the template (`{{if .Owner}}`) would be prudent.
> 
> _Written by Claude_

All of these should be resolved.

**Review by silverwind — APPROVED** [review]

File: templates/package/shared/cleanup_rules/preview.tmpl [hunk 3]
```diff
@@ -18,7 +18,14 @@
...
 				<tr>
 					<td>{{.Package.Type.Name}}</td>
-					<td>{{.Package.Name}}</td>
+					<td>
+						{{.Package.Name}}
+						{{if .Owner}}
+							<span class="label-list">
+								{{template "package/shared/visibility_badge" dict "Package" .Package "Owner" .Owner}}
+							</span>
+						{{end}}
```
Comment (wxiaoguang) [inline templates/package/shared/cleanup_rules/preview.tmpl]: I don't think this `if` check is right or necessary.

Show me the real case why it is needed.

File: templates/package/settings.tmpl [hunk 4]
```diff
@@ -10,12 +10,32 @@
...
+			<a href="{{.PackageDescriptor.PackageWebLink}}">{{.PackageDescriptor.Package.Name}}</a>
+			<span class="label-list">
+				{{template "package/shared/visibility_badge" dict "Package" .PackageDescriptor.Package "Owner" .PackageDescriptor.Owner}}
+			</span>
+			/ <strong>{{ctx.Locale.Tr "repo.settings"}}</strong>
+		</p>
+		{{$visibilitySettingsLink := print AppSubUrl "/user/settings"}}
+		{{if .ContextUser.IsOrganization}}
+			{{$visibilitySettingsLink = print .ContextUser.OrganisationLink "/settings"}}
+		{{end}}
```
Comment (wxiaoguang) [inline templates/package/settings.tmpl]: No. No more template logic


Image

File: templates/package/shared/cleanup_rules/preview.tmpl (hunk 3 above)
Comment (lunny) [reply templates/package/shared/cleanup_rules/preview.tmpl]: 7986adca01

File: templates/package/settings.tmpl (hunk 4 above)
Comment (lunny) [reply templates/package/settings.tmpl]: 7986adca01

File: templates/package/shared/cleanup_rules/preview.tmpl (hunk 3 above)
Comment (wxiaoguang) [reply templates/package/shared/cleanup_rules/preview.tmpl]: If no need to add the "if", why you ever changed it?

You are just the slave of AI? Or does @bircni know why the AI review make sense?

image

**Review by Zettat123 — APPROVED** [review]

**Comment by silverwind** [comment]
@lunny first resolve all open review threads before you attempt merging, https://github.com/go-gitea/gitea/pull/37610#discussion_r3214396154 is still uncommented and open.

And stop setting `reviewed/wait-merge` without also enabling auto-merge, this just burns massive amount of CI minutes for no good reason, you just wasted ~400 CI minutes. That label is only for when multiple PRs are awaiting merge.

**Comment by silverwind** [comment]
Merging with open review comments is such bad practice, I should have retracted my review.

**Comment by lunny** [comment]
> @lunny first resolve all open review threads before you attempt merging, [(earlier comment)](https://github.com/go-gitea/gitea/pull/37610#discussion_r3214396154) is still uncommented and open.
> 
> And stop setting `reviewed/wait-merge` without also enabling auto-merge, this just burns massive amount of CI minutes for no good reason, you just wasted ~400 CI minutes. That label is only for when multiple PRs are awaiting merge.

That has been resolved. @wxiaoguang mentioned that the AI review was incorrect, but I followed the feedback accordingly.

**Comment by silverwind** [comment]
Ok, then "Resolve" and optionally comment so that is clear.
